File: file/src/main/scala/farjs/file/FileViewHistory.scala
package farjs.file

import farjs.filelist.history.{History, HistoryKind}

import scala.scalajs.js
import scala.scalajs.js.annotation.JSImport

sealed trait FileViewHistory extends js.Object {
  val path: String
  val params: FileViewHistoryParams
}

@js.native
@JSImport("../file/FileViewHistory.mjs", JSImport.Default)
object NativeFileViewHistory extends js.Object {

  val fileViewsHistoryKind: HistoryKind = js.native

  def toHistory(h: FileViewHistory): History = js.native

  def fromHistory(h: History): js.UndefOr[FileViewHistory] = js.native

  def pathToItem(path: String, isEdit: Boolean): String = js.native
}

object FileViewHistory {

  def apply(path: String, params: FileViewHistoryParams): FileViewHistory = {
    js.Dynamic.literal(
      path = path,
      params = params
    ).asInstanceOf[FileViewHistory]
  }

  def unapply(arg: FileViewHistory): Option[(String, FileViewHistoryParams)] = {
    Some((
      arg.path,
      arg.params
    ))
  }
  
  val fileViewsHistoryKind: HistoryKind = NativeFileViewHistory.fileViewsHistoryKind
  
  def toHistory(h: FileViewHistory): History =
    NativeFileViewHistory.toHistory(h)

  def fromHistory(h: History): js.UndefOr[FileViewHistory] =
    NativeFileViewHistory.fromHistory(h)
  
  def pathToItem(path: String, isEdit: Boolean): String =
    NativeFileViewHistory.pathToItem(path, isEdit)
}

File: file/FileViewHistory.mjs
/**
 * @typedef {import("@farjs/filelist/history/HistoryProvider.mjs").History} History
 * @typedef {import("@farjs/filelist/history/HistoryProvider.mjs").HistoryKind} HistoryKind
 */
import { stripPrefix } from "@farjs/filelist/utils.mjs";

/**
 * @typedef {{
 *  readonly isEdit: boolean;
 *  readonly encoding: string;
 *  readonly position: number;
 *  readonly wrap?: boolean;
 *  readonly column?: number;
 * }} FileViewHistoryParams
 */

/**
 * @typedef {{
 *  readonly path: string;
 *  readonly params: FileViewHistoryParams;
 * }} FileViewHistory
 */

/**
 * @param {string} path
 * @param {FileViewHistoryParams} params
 * @returns {FileViewHistory}
 */
function FileViewHistory(path, params) {
  return { path, params };
}

/** @type {HistoryKind} */
FileViewHistory.fileViewsHistoryKind = {
  name: "farjs.fileViews",
  maxItemsCount: 150,
};

/**
 * @param {FileViewHistory} h
 * @returns {History}
 */
FileViewHistory.toHistory = (h) => {
  return {
    item: FileViewHistory.pathToItem(h.path, h.params.isEdit),
    params: h.params,
  };
};

/**
 * @param {History} h
 * @returns {FileViewHistory | undefined}
 */
FileViewHistory.fromHistory = (h) => {
  if (h.params) {
    return {
      path: FileViewHistory._itemToPath(h.item),
      params: /** @type {FileViewHistoryParams} */ (h.params),
    };
  }
  return undefined;
};

/**
 * @param {string} path
 * @param {boolean} isEdit
 * @returns {string}
 */
FileViewHistory.pathToItem = (path, isEdit) => {
  return isEdit ? `E:${path}` : `V:${path}`;
};

/**
 * @param {string} item
 * @returns {string}
 */
FileViewHistory._itemToPath = (item) => {
  if (item.startsWith("V:")) return stripPrefix(item, "V:");
  else if (item.startsWith("E:")) return stripPrefix(item, "E:");
  else return item;
};

export default FileViewHistory;
